File: Graph Algorithms/Word ladder problem (BFS).cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
#define endl "\n"

const int N = 100; 
const int mod = 1e9 + 7;

int n, m, k;

// bool isOneDiff(string a, string b)
// {
//     if (a.length() != b.length()) 
//         return false;
    
//     int diff = 0;
//     for (int i = 0; i < a.length(); i++) 
//     {
//         if (a[i] != b[i]) 
//             diff++;
//         if (diff > 1) 
//             return false;
//     }

//     if (diff == 1)  return true;
//     return false;
// }

int ladderLength(string beginWord, string endWord, set <string>& dict)
{
    if ( dict.find(endWord) == dict.end())
        return 0;

    int level = 0, wordlength = beginWord.size();

    queue <string> frontier;
    frontier.push(beginWord);

    while(! frontier.empty())
    {
        level++;
        int size = frontier.size();
        for (int i = 0; i < size; i++)
        {
            string currWord = frontier.front();
            frontier.pop();

            for (int pos = 0; pos < wordlength; pos++)
            {
                char orig_char = currWord[pos];
                for (char c = 'a'; c <= 'z'; c++)
                {
                    currWord[pos] = c;
                    if (currWord == endWord) 
                        return level + 1;
                    if (dict.find(currWord) == dict.end())
                        continue;
                    dict.erase(currWord);
                    frontier.push(currWord);
                }
                currWord[pos] = orig_char;
            }

            // for (auto itr = dict.begin(); itr != dict.end(); itr++)
            // {
            //     if (isOneDiff(currWord, *itr)) 
            //     {
            //         if (*itr == endWord) 
            //             return level + 1;
            //         cout << *itr << " ";
            //         string w = *itr;
            //         frontier.push(w);
            //         dict.erase(w);
            //     }
            // }
        }    
    }

    return 0;
}

void test_case()
{
    set<string> dict = {"cool", "pool", "poll", "foil", "fail", "pole", "pope", "pale", "page", "sage", "sale", "fall"};
    // dict.insert("hit");
    // dict.insert("hot");
    // dict.insert("cot");
    // dict.insert("dog");
    // dict.insert("dot");
    // dict.insert("log");
    // dict.insert("cog");
    string start = "fool", end = "sage";
    cout << ladderLength(start, end, dict);

    cout << endl;
    return;
}


































void solve()
{
    int t;
    t = 1;
    // cin >> t;
    for (int nr = 1; nr <= t; nr++)
    {
        test_case();
    }
}

int main()
{   
    // freopen("input.txt", "r", stdin);
    // freopen("output.txt", "w", stdout);
    //synchronize c++ streams with c streams
    ios_base::sync_with_stdio(false);
    //guarentes flushing
    cin.tie(NULL);
    cout.tie(NULL);
    solve();
}

File: Graph Algorithms/Word ladder problem (BFS).py
from queue import Queue
from string import ascii_lowercase

def shortestPath(start, target, wordDict):
    frontier = Queue()
    frontier.put(start)

    level = 0
    visited = []

    while frontier.empty() == False:  # O(M) because word changes by one character
        level += 1
        size = frontier.qsize()   

        for _ in range(size):         #for this level # O(N)
            curr = frontier.get()
            visited.append(curr)

            if curr == target:       # if current is target
                return level

            for pos in range(len(curr)):     # O(M)* 26 check all possible neighbors
                tmp = curr
                for ch in ascii_lowercase:
                    curr = list(curr)  # string to list
                    curr[i] = ch        # push 
                    curr = ''.join(curr) # list to string
                    if curr not in visited and curr in wordDict:  # if not visited and is in wordDict
                        frontier.put(curr) 
                curr = tmp

            # time complexity O(N * M)
            # for i in range(len(wordDict)):  #check all possible neigbors 
            #     if oneDiff(curr, wordDict[i]) and wordDict[i] not in visited:
            #         frontier.put(wordDict[i]) 
    
    return 0
                

if __name__ == "__main__":
    start = "fool"
    target = "sage"
    wordDict = ["cool", "pool", "poll", "foil", "fail", "pole", "pope", "pale", "page", "sage", "sale", "fall"]
    
    print(shortestPath(start, target, wordDict))
